torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -125,9 +132,62 @@ class ProcessGroupNCCL : public ProcessGroup {
     // to the store.
     std::shared_ptr<Store> store_;
 
+    // Store a Future work associated with WorkNCCL.
+    c10::intrusive_ptr<c10::ivalue::Future> futureWork_;
+
     friend class ProcessGroupNCCL;
   };
 
+  // FutureNCCL is a subclass of c10d ivalue's Future. FutureNCCL has a
+  // reference to WorkNCCL and NCCL collective's outputs. The goal is to
+  // use this class in getFuture API of WorkNCCL. Once NCCL kernel is done
+  // executing, this future will have NCCL collective's outputs as its value.
+  // It overrides wait(), markCompleted(), and addCallback() methods of Future.
+  struct FutureNCCL : at::ivalue::Future {
+   public:
+    explicit FutureNCCL(
+        std::shared_ptr<ProcessGroupNCCL::WorkNCCL> work,
+        std::vector<at::Tensor>& outputs)
+        : at::ivalue::Future(c10::ListType::create(c10::TensorType::get())),
+          work_(work),
+          outputs_(std::make_shared<std::vector<at::Tensor>>(outputs)) {}
+
+    // Simply calls WorkNCCL's wait(). After WorkNCCL's wait() returns it calls
+    // markCompleted using collective operation's outputs. It will return after
+    // synchronizing the correct GPU streams to ensure we can have async CUDA
+    // execution and it does not wait for the entire operation to complete on
+    // GPU. If NCCL_BLOCKING_WAIT is enabled, in that case, it will wait for the
+    // entire operation to complete before returning.
+    void wait() override {
+      work_->wait();
+      markCompleted(at::IValue(*outputs_));
```

Comment:
The issue here is that calling `completed()` on this Future will return True despite the work on the GPU not finishing. I think we should do the following:

1. Don't set `completed_ = true`.
2. Override `markCompleted()` and ensure it throws an Exception mentioning that the operation is not supported.
3. Override `value` and `constValue` to return `outputs_`.
4. Override `hasValue` to also return true.
5. Override `completed()` to return `work_->isCompleted()`.